### PR TITLE
Add additional-capability parameter to gnmi yang models

### PR DESCRIPTION
**Pull request by caladi**
Introduce the ability to add additional capability to node
 schema context. Some devices not provide complete
 capability list, so missing capabilities must
 be added by this addtional-capability list parameter.
Add tests for cover this functionality.
Add readme for describing this functionality.

Signed-off-by: Ivan Caladi <ivan.caladi@pantheon.tech>

### CODE REVIEW

File: lighty-modules/lighty-gnmi/lighty-gnmi-test/src/test/java/io/lighty/modules/gnmi/test/gnmi/rcgnmi/GnmiConnectionITTest.java [hunk 1]
```diff
@@ -513,4 +539,16 @@ public void connectDeviceWithIncorrectCredentialsTest()
...
+        assertEquals(HttpURLConnection.HTTP_OK, getGnmiTopologyResponse.statusCode());
+        final JSONArray topologies =
+            new JSONObject(getGnmiTopologyResponse.body()).getJSONArray("network-topology:topology");
+        assertEquals(1, topologies.length());
+        final JSONObject gnmiTopologyJSON = topologies.getJSONObject(0);
+        LOG.info("Empty gnmi-topology check response: {}", gnmiTopologyJSON);
+        assertEquals("gnmi-topology", gnmiTopologyJSON.getString("topology-id"));
+        assertThrows(JSONException.class, () -> gnmiTopologyJSON.getJSONArray("node"));
+    }
+
```
Comment (martinbugan-pt) [inline lighty-modules/lighty-gnmi/lighty-gnmi-test/src/test/java/io/lighty/modules/gnmi/test/gnmi/rcgnmi/GnmiConnectionITTest.java:553]: You should not change something that is not described in commit message - also tests are written to be verbose and were purposely not packaged into methods (but if you want to do that, methods should not contain asserts, they should be visible in tests and it should be clear what is asserting as output from the method), so please revert this changes with this method

**Review by martinbugan-pt — CHANGES_REQUESTED** [review]

File: lighty-models/lighty-gnmi-models/lighty-gnmi-yang-storage-model/src/main/yang/gnmi-yang-storage.yang [hunk 2]
```diff
@@ -31,21 +35,44 @@ module gnmi-yang-storage {
...
-        }
+        uses yang-model-name-version;
         leaf body {
             description "Body of the yang model";
             type string;
         }
     }
 
+    grouping additional-yang-models {
+        list additional-capability {
```
Comment (marekzatko) [inline lighty-models/lighty-gnmi-models/lighty-gnmi-yang-storage-model/src/main/yang/gnmi-yang-storage.yang:58]: consider using leaf-list, since we do not need to identifiy specific yang model (we simply add all additional models to capability response)

**Review by marekzatko — CHANGES_REQUESTED** [review]

**Review by marekzatko — APPROVED** [review]
I personally like your refactored test more. That extracted method can easily be named and is obvious on first glance on what it does. This verbose style does not add any  value  in readability. But if others disagree, let it be this way.

**Review by martinbugan-pt — APPROVED** [review]

File: lighty-modules/lighty-gnmi/lighty-gnmi-sb/src/main/java/io/lighty/gnmi/southbound/device/connection/DeviceConnectionManager.java [hunk 3]
```diff
@@ -125,6 +130,25 @@ Establish connection with device (future will be set with GnmiDeviceManager for
             executorService);
     }
 
+    private Gnmi.CapabilityResponse getExtensionYangModelsList(final DeviceConnection deviceConnection) {
+        Optional<Map<AdditionalCapabilityKey, AdditionalCapability>> models =
+            deviceConnection.getConfigurableParameters().getAdditionalCapabilities();
+        Gnmi.CapabilityResponse.Builder capabilitiesResponseBuilder = Gnmi.CapabilityResponse.newBuilder();
```
Comment (vainyksi) [inline lighty-modules/lighty-gnmi/lighty-gnmi-sb/src/main/java/io/lighty/gnmi/southbound/device/connection/DeviceConnectionManager.java:136]: make `capabilitiesResponseBuilder` final

File: lighty-modules/lighty-gnmi/lighty-gnmi-sb/src/main/java/io/lighty/gnmi/southbound/device/connection/DeviceConnectionManager.java [hunk 4]
```diff
@@ -125,6 +130,25 @@ Establish connection with device (future will be set with GnmiDeviceManager for
             executorService);
     }
 
+    private Gnmi.CapabilityResponse getExtensionYangModelsList(final DeviceConnection deviceConnection) {
```
Comment (vainyksi) [inline lighty-modules/lighty-gnmi/lighty-gnmi-sb/src/main/java/io/lighty/gnmi/southbound/device/connection/DeviceConnectionManager.java:133]: rename `getExtensionYangModelsList` ... -> `getExtensionCapabilitiesList`
you are working with capabilities in method, and return type of method is also capability, mixing of names may confuse reader

File: lighty-modules/lighty-gnmi/lighty-gnmi-sb/src/main/java/io/lighty/gnmi/southbound/device/connection/DeviceConnectionManager.java [hunk 5]
```diff
@@ -125,6 +130,25 @@ Establish connection with device (future will be set with GnmiDeviceManager for
...
+        Optional<Map<AdditionalCapabilityKey, AdditionalCapability>> models =
+            deviceConnection.getConfigurableParameters().getAdditionalCapabilities();
+        Gnmi.CapabilityResponse.Builder capabilitiesResponseBuilder = Gnmi.CapabilityResponse.newBuilder();
+        if (models.isPresent()) {
+            models.get().entrySet()
+                .stream()
+                .map(model -> model.getValue())
+                .peek(additionalCapability ->
+                    LOG.debug("Adding additional gNMI capability: name {} version {}",
```
Comment (vainyksi) [inline lighty-modules/lighty-gnmi/lighty-gnmi-sb/src/main/java/io/lighty/gnmi/southbound/device/connection/DeviceConnectionManager.java:142]: this is not actually the adding of capability to capabilities list used for node, it should not be stated like this before it was actually done, there are multiple places where it can fail after this log, so the capability may not be added actually

File: lighty-modules/lighty-gnmi/lighty-gnmi-test/src/test/java/io/lighty/modules/gnmi/test/gnmi/rcgnmi/GnmiConnectionITTest.java [hunk 6]
```diff
@@ -203,6 +209,87 @@ public void connectDeviceCorrectlyTest()
...
+    @Test
+    public void connectDeviceWithAdditionalCapabilityAndModelTest()
+        throws InterruptedException, IOException, ExecutionException, TimeoutException {
+        final HttpResponse<String> getGnmiTopologyResponse = sendGetRequestJSON(GNMI_TOPOLOGY_PATH);
+        assertEquals(HttpURLConnection.HTTP_OK, getGnmiTopologyResponse.statusCode());
+        final JSONArray topologies =
+            new JSONObject(getGnmiTopologyResponse.body()).getJSONArray("network-topology:topology");
+        assertEquals(1, topologies.length());
+        final JSONObject gnmiTopologyJSON = topologies.getJSONObject(0);
+        LOG.info("Empty gnmi-topology check response: {}", gnmiTopologyJSON);
```
Comment (vainyksi) [inline lighty-modules/lighty-gnmi/lighty-gnmi-test/src/test/java/io/lighty/modules/gnmi/test/gnmi/rcgnmi/GnmiConnectionITTest.java:221]: remove the log or rather use the message parameter in `assertThrows` (as this log relates to that assertion that there are no nodes)

**Review by vainyksi — CHANGES_REQUESTED** [review]
